### PR TITLE
test: test_should_not_delegate_funds_if_already_delegated

### PR DESCRIPTION
**Pull request by netpoe**

### CODE REVIEW

File: rust-escrow/conditional-escrow/src/lib.rs [hunk 1]
```diff
@@ -171,19 +172,32 @@ impl ConditionalEscrow {
...
+            PromiseResult::Successful(result) => {
+                let res = String::from_utf8_lossy(&result);
+                if res == "true" {
+                    self.total_funds = 0;
+                    return true;
+                }
+                false
+            }
+            _ => false,
+        }
```
Comment (netpoe) [inline rust-escrow/conditional-escrow/src/lib.rs:198]: Would this implementation work the same as this one? I think it is cleaner:
https://github.com/aufacicenta/near.holdings/blob/master/rust-escrow/src/lib.rs#L183-L199

```suggestion
        pub fn on_create_dao_callback(&mut self) -> bool {
        assert_eq!(env::promise_results_count(), 1, "ERR_CALLBACK_METHOD");

        if near_sdk::is_promise_success() {
            self.total_funds = 0;
            true
        } else {
            false
        }
    }
```

File: rust-escrow/dao-factory/src/lib.rs [hunk 2]
```diff
@@ -91,21 +91,34 @@ impl DaoFactory {
...
-                    .insert(&(country_code), &(daos_by_country));
-                true
+            PromiseResult::Successful(result) => {
+                let res = String::from_utf8_lossy(&result);
+                if res == "true" {
+                    self.daos_by_country_count.insert(&(country_code), &(daos_by_country));
+                    return true;
+                }
+                Promise::new(predecessor_account_id).transfer(attached_deposit);
+                false
```
Comment (netpoe) [inline rust-escrow/dao-factory/src/lib.rs:121]: Same as above.
Perhaps this implementation is cleaner and provides same result: https://github.com/aufacicenta/near.holdings/blob/master/rust-escrow/src/lib.rs#L183-L199

**Comment by netpoe** [comment]
@susguzman 
Let's add a new arg to `ConditionalEscrow::new`: `dao_account_id`
And a new method: `set_dao_account_id`, in case the creation fails upon `delegate_funds`, we are able to try with a new name.

File: rust-escrow/conditional-escrow/src/lib.rs [hunk 3]
```diff
@@ -549,7 +549,116 @@ mod tests {
...
+        );
+
+        contract.delegate_funds("dao1".to_string());
+
+        testing_env!(
+            context.build(),
+            near_sdk::VMConfig::test(),
+            near_sdk::RuntimeFeesConfig::test(),
+            Default::default(),
+            vec![PromiseResult::Successful("true".to_string().into_bytes())],
```
Comment (netpoe) [inline rust-escrow/conditional-escrow/src/lib.rs:601]: Nice!

I was thinking, even that SputnikDAO is returning false. What if we in this contract, actually panicks. To be able to use the other `is_promise_successful` pattern?

File: rust-escrow/dao-factory/src/lib.rs [hunk 4]
```diff
@@ -89,23 +83,39 @@ impl DaoFactory {
...
+                    let dao_account_id: AccountId = format!("{}.{}", dao_name, env::current_account_id()).parse().unwrap();
+                    self.dao_index
+                        .insert(&predecessor_account_id, &dao_account_id);
+
+                    return true;
+                }
+
+                Promise::new(predecessor_account_id).transfer(attached_deposit);
+
+                false
```
Comment (netpoe) [inline rust-escrow/dao-factory/src/lib.rs:118]: Same as above. What if we actually panick! instead of returning `false`?

**Review by netpoe — COMMENTED** [review]
Much simpler 👍 , it feels better. 
Let me know wdyt about my comments regarding our contracts panicking instead of returning `false` too.

File: rust-escrow/dao-factory/src/lib.rs (hunk 4 above)
Comment (susguzman) [reply rust-escrow/dao-factory/src/lib.rs:118]: Yes, I think that is better instead a bool. We can catch the response easier.

**Review by susguzman — APPROVED** [review]

**Pull request merged**